### PR TITLE
Dictionary - parameterised tests refactoring

### PR DESCRIPTION
**Pull request by edualonso**
- Removed leftover support for boxed byte arrays in RealmMapEntrySet.
- Changed the approach to testing dictionaries. Now we have parameterised tests which take into consideration all supported types for dictionaries, also for `Mixed`.
- Only two classes, `UnmanagedGenericTester` and `ManagedGenericTester`, are used to test all supported types. In the case of `ManagedGenericTester` Kotlin's `KFunction1` and `KFunction2` are used to ease working with field getters/setters to standardise testing under the umbrella of just one class.
- Tests have a harmonised way to assert conditions whenever type constraints make it more difficult via a `TypeAsserter`. E.g. one cannot compare an unmanaged `RealmModel` to a managed one - same happens with `Mixed` instances wrapping `RealmModel`s.

### CODE REVIEW

**Comment by edualonso** [comment]
CI fails due failing stress tests. It seems like it's due to a bug in core: https://github.com/realm/realm-core/issues/4465

File: realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    UNMANAGED, MANAGED
+}
+
+/**
+ * Supported types by dictionaries. Notice that Mixed dictionaries can in turn support all these
+ * types internally (except Mixed itself).
+ *
+ * Add new types ad-hoc here.
+ */
+enum class DictionarySupportedType {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt:188]: I think we should rather iterate our tests over an _all possible types_ enumeration and define a utility method `isDictionarySupported` (possible on the enum itself) that returns true for the below similar ones. Then it is way easier to make tests that fails it we introduce a new type.

File: realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+ * [RealmDictionary] types (i.e. all primitive Realm types (see [DictionarySupportedType]) plus
+ * [RealmModel] and [Mixed] (and in turn all possible types supported by Mixed) in both `managed`
+ * and `unmanaged` modes.
+ *
+ * In order to streamline the testing for managed dictionaries we use Kotlin's reflection API
+ * `KFunction1` and `KFunction2`. These two methods provide access to the Java getters and setters
+ * used to work with each dictionary field.
+ */
+@RunWith(Parameterized::class)
+class ParameterizedDictionaryTests(
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt:43]: When you have this pattern of almost delegating directly through to a specific incoming parameter you can use inheritance by delegation to automatically. Don't know now to put annotations on the individual methods without explicitly defining them though 😞

File: realm/realm-library/src/main/java/io/realm/ManagedMapManager.java [hunk 3]
```diff
@@ -335,6 +335,9 @@ public Mixed put(Object key, @Nullable Mixed value) {
 
     @Override
     public boolean containsValue(@Nullable Object value) {
+        if (value == null) {
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/ManagedMapManager.java:338]: Sounds like we are not able to hold `null`s. Is this documented properly somewhere?

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTester.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm
+
+/**
+ * Contains all the methods to test in RealmDictionary plus some other convenience ones.
+ */
+interface DictionaryTester {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTester.kt:22]: I give that it is on a tester interface, but the names are a bit bean-ish, so you get a bit confused when reading the implementations. Probably ok to keep.

**Review by rorbech — APPROVED** [review]
Some minor comments. I am not that much into the overall dictionary implementation, so mostly a _code sanity_ review, but seems like an overall good approach to ensure coverage on these generic types. A bit afraid of yet another enum `DictionarySupportedType`. Maybe we could do a central enum for test types with properties for `isDictSupported` etc.

File: realm/realm-library/src/testUtils/java/io/realm/entities/AllTypes.java [hunk 5]
```diff
@@ -60,10 +61,31 @@
...
+    public static final String FIELD_OBJECT_ID_DICTIONARY = "columnObjectIdDictionary";
+    public static final String FIELD_UUID_DICTIONARY = "columnUUIDDictionary";
+    public static final String FIELD_DECIMAL128_DICTIONARY = "columnDecimal128Dictionary";
+    public static final String FIELD_MIXED_DICTIONARY = "columnMixedDictionary";
+
     public static final String[] INVALID_TYPES_FIELDS_FOR_DISTINCT
             = new String[] {FIELD_REALMOBJECT, FIELD_REALMLIST, FIELD_DOUBLE, FIELD_FLOAT,
             FIELD_STRING_LIST, FIELD_BINARY_LIST, FIELD_BOOLEAN_LIST, FIELD_LONG_LIST,
-            FIELD_DOUBLE_LIST, FIELD_FLOAT_LIST, FIELD_DATE_LIST};
+            FIELD_DOUBLE_LIST, FIELD_FLOAT_LIST, FIELD_DATE_LIST, FIELD_BOOLEAN_DICTIONARY,
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/testUtils/java/io/realm/entities/AllTypes.java:82]: I agree, but it isn't feasible for me to do it now as a part of the dictionaries feature. I have added a comment to https://github.com/realm/realm-java/issues/7320#issuecomment-790053382

File: realm/realm-library/src/main/java/io/realm/ManagedMapManager.java (hunk 3 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/ManagedMapManager.java:338]: The way to store null values in Mixed is by using `Mixed.nullValue()`. I think it should be documented somewhere in one of the mixed classes. I will write a side note here though.

File: realm/realm-library/src/testUtils/java/io/realm/entities/AllTypes.java [hunk 6]
```diff
@@ -106,6 +128,23 @@
     private RealmList<UUID> columnUUIDList;
     private RealmList<Mixed> columnMixedList;
 
+    private RealmDictionary<DogPrimaryKey> columnRealmDictionary;
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/testUtils/java/io/realm/entities/AllTypes.java:131]: They are, but it was more convenient for me to test it like this. Funnily enough this has uncovered a bug in the thread stress test.

File: realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt (hunk 1 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/ParameterizedDictionaryTests.kt:188]: I have created https://github.com/realm/realm-java/issues/7331

**Pull request merged**